### PR TITLE
Add contracts for user registration

### LINKED ISSUE #3: Add Semaphore contract gatekept by HatsExcubia
- [x] Add semaphore contract
- [x] Add excubiae
- [x] Deployment script

### PR DESCRIPTION
**Pull request by 0xjei**
This PR adds everything needed to register users with a gatekeep mechanism based on `hats` protocol using smart contracts

- [x] basic configuration for `foundry` with npm packages
- [x] minimal `Semaphore` contract
- [x] gatekeep w/ `SemaphoreExcubia`
- [x] script x deploy

### CODE REVIEW

File: packages/contracts/script/Hackathon.s.sol [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (ctrlc03) [inline packages/contracts/script/Hackathon.s.sol]: MIT?

File: packages/contracts/script/Hackathon.s.sol [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import { Semaphore } from "../src/Semaphore.sol";
+
+contract Deploy is Script {
+    // @todo needs the correct HatsProtocol address.
+    // See https://book.getfoundry.sh/tutorials/solidity-scripting for deploy.
+    address public HATS_ADDRESS = address(0x3bc1A0Ad72417f2d411118085256fC53CBdDd137);
+    uint256[] public CRITERION_HATS = [0, 1, 2];
+
+    function run() external {
+        // @todo uncomment this.
```
Comment (ctrlc03) [inline packages/contracts/script/Hackathon.s.sol]: uncomment which line?

File: packages/contracts/src/Semaphore.sol [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+import { ISemaphore } from "@semaphore-protocol/contracts/interfaces/ISemaphore.sol";
+import { ISemaphoreVerifier } from "@semaphore-protocol/contracts/interfaces/ISemaphoreVerifier.sol";
+import { SemaphoreGroups } from "@semaphore-protocol/contracts/base/SemaphoreGroups.sol";
+import { MIN_DEPTH, MAX_DEPTH } from "@semaphore-protocol/contracts/base/Constants.sol";
+import { HatsExcubia } from "@zk-kit/excubiae/extensions/HatsExcubia.sol";
+
+/// @title Semaphore
+/// @dev A minimal version of the Semaphore contract.
+contract Semaphore is ISemaphore, SemaphoreGroups {
```
Comment (ctrlc03) [inline packages/contracts/src/Semaphore.sol:12]: is this pattern better than extending Semaphore and overriding the functions we changed?

**Review by ctrlc03 — COMMENTED** [review]
Thanks, left some comments

File: packages/contracts/src/Semaphore.sol (hunk 3 above)
Comment (0xjei) [reply packages/contracts/src/Semaphore.sol:12]: `Semaphore.sol` functions are not virtual and we are just taking Semaphore and extrapolating what we need to handle one group. This should cost less gas and be resilient enough for the sake of the hackathon.

Comment (ctrlc03) [reply packages/contracts/src/Semaphore.sol:12]: perfect

**Review by ctrlc03 — APPROVED** [review]
Thanks

**Pull request merged**